### PR TITLE
Gerrit: fix server version parsing

### LINKED ISSUE #7521: [NBug] Version string portion was too short or too long.
:warning: The sections below must be filled in and this text must be removed or the issue will be closed.

## Error Details
```
System.ArgumentException: Version string portion was too short or too long.
   at System.Version.VersionResult.SetFailure(ParseFailureKind failure, String argument)
   at System.Version.TryParseVersion(String version, VersionResult& result)
   at System.Version.Parse(String input)
   at Gerrit.FormGerritBase.OnLoad(EventArgs e)
   at Gerrit.FormGerritPublish.OnLoad(EventArgs e)
   at System.Windows.Forms.Form.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at ResourceManager.GitExtensionsFormBase.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

## Additional information
Trying to publish to Gerrit

## Environment
- Git Extensions 3.3.0.7719
- Build e52f5178e8d3c9899efbe6fae66026865afa3fe1
- Git 2.24.1.windows.2
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4075.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by dfev77**
Do not parse version if request to retrieve it was unsuccessful

Fixes #7521

This is just a quick fix, see notes below.

### After

Currently GitExt makes the following incorrect assessments about the Gerrit server and further changes are required to have a proper Gerrit integration (separate topic):
 - server is available under https if protocol not specified
 - port is the default protocol configuration

In our case ".gerrit" configuration looks like 
`[gerrit]
host=repo.example.lan
port=1234
defaultremote=origin
project=reponame`
while the gerrit is available at "http://repo.example.lan:8080"

## Test methodology

- Manually tested

## Test environment(s)

- GIT 2.23.0
- Windows 10
- Gerrit 2.15.3-1-g047eac59e7-dirty
:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by lekhmanrus** [comment]
LGTM

File: Plugins/Gerrit/FormGerritBase.cs [hunk 1]
```diff
@@ -66,7 +66,15 @@ private Version GetGerritVersion()
         {
             RestRequest request = new RestRequest("/config/server/version");
             IRestResponse response = client.Execute(request);
-            return Version.Parse(response.Content.Replace(")]}'", "").Replace("\n", "").Replace("\"", "").Split('-')[0]);
+            if (response.IsSuccessful)
+            {
+                string versionText = response.Content.Replace(")]}'", "").Replace("\n", "").Replace("\"", "").Split('-')[0];
+                return Version.Parse(versionText);
```
Comment (RussKie) [inline Plugins/Gerrit/FormGerritBase.cs]: Wouldn't it be easier to replace this with a regex?
Something like \d{1,3}\.\d{1,3}(\.\d{1,3}){0,2}

Comment (RussKie) [reply Plugins/Gerrit/FormGerritBase.cs]: GitHub's eaten escaping, dot must be escaped too

File: Plugins/Gerrit/FormGerritBase.cs [hunk 2]
```diff
@@ -66,7 +66,15 @@ private Version GetGerritVersion()
         {
             RestRequest request = new RestRequest("/config/server/version");
             IRestResponse response = client.Execute(request);
-            return Version.Parse(response.Content.Replace(")]}'", "").Replace("\n", "").Replace("\"", "").Split('-')[0]);
+            if (response.IsSuccessful)
```
Comment (RussKie) [inline Plugins/Gerrit/FormGerritBase.cs]: Please invert `if` to reduce nesting

File: Plugins/Gerrit/FormGerritBase.cs (hunk 1 above)
Comment (RussKie) [inline Plugins/Gerrit/FormGerritBase.cs]: Unless `versionText` conforms to a string parseable to `Version` this will fail again.

This is where a regex may help, because you can enforce the format (though not perfectly): https://docs.microsoft.com/en-us/dotnet/api/system.version.-ctor?view=netframework-4.8

```^(\d*\.\d*(\.\d*){0,2})$```
image
Here I am obviously testing for a string from a beginning and to the end of the line, given I have no idea `/config/server/version` return payload is... 

However `Version.Parse(string)` call must still be wrapped in `try..catch` https://docs.microsoft.com/en-us/dotnet/api/system.version.parse?view=netframework-4.8


So the method should look something like the following:

```cs
            RestRequest request = new RestRequest("/config/server/version");
            IRestResponse response = client.Execute(request);
            if (!response.IsSuccessful)
			{
				return GerritUtil.GerritWithPrivateReviewSupport;
			}

			try
			{
				var m = Regex.Match(..., ...);
				if (m.Success)
				{
					return Version.Parse(m....);
				}
			}
			catch
			{
				// no-op
			}

			return GerritUtil.GerritWithPrivateReviewSupport;
```

I would also encourage you to write tests to test edge-cases.

File: Plugins/Gerrit/FormGerritBase.cs (hunk 2 above)
Comment (dfev77) [reply Plugins/Gerrit/FormGerritBase.cs]: replaced it with ternary operator

File: Plugins/Gerrit/FormGerritBase.cs (hunk 1 above)
Comment (dfev77) [reply Plugins/Gerrit/FormGerritBase.cs]: changed to regex matching

File: Plugins/Gerrit/FormGerritBase.cs [hunk 3]
```diff
@@ -66,7 +66,17 @@ private Version GetGerritVersion()
         {
             RestRequest request = new RestRequest("/config/server/version");
             IRestResponse response = client.Execute(request);
-            return Version.Parse(response.Content.Replace(")]}'", "").Replace("\n", "").Replace("\"", "").Split('-')[0]);
+
+            return (response.IsSuccessful
+                && Version.TryParse(ExtractVersion(response.Content), out var version))
+                ? version : GerritUtil.GerritWithPrivateReviewSupport;
```
Comment (RussKie) [inline Plugins/Gerrit/FormGerritBase.cs]: We still need to wrap it in `try..catch`, otherwise we risk crashing the app again, if something is wrong.

Comment (dfev77) [reply Plugins/Gerrit/FormGerritBase.cs]: Looking at possible error sources:
 - rest request handles exceptions - see https://github.com/restsharp/RestSharp/blob/master/RestSharp/RestClient.Sync.cs
 - Version.TryParse does not throw
 - Regex.Match might throw if input is null or regex is invalid.

 However in this last case, as long as we intend to handle it, should we just log it or just show a message box? (need some guidance on the error handling mechanism on GitEx)

File: Plugins/Gerrit/FormGerritBase.cs (hunk 1 above)
Comment (dfev77) [reply Plugins/Gerrit/FormGerritBase.cs]: @RussKie on testing, the only feasible way would be to extract the version retrieval part to a separate class and mock the RestClient. (
`class GerritServer { 
public GerritServer(IRestClient client, GerritSettings settings)
{...}
public Version GetVersion();
{...}
}`
Then the GerritSettings class would also need some changes  as it not test friendly (cannot be stubbed and requires file parsing under a IGitModule context)

So, unit tests would be indeed nice, but right now seem a little to hard to write. So it will take some time....

**Comment by RussKie** [comment]
My apologies, I missed that it was TryParse. You're correct, a chance of an
exception should be low.
All good then!

File: Plugins/Gerrit/Server/GerritCapabilities.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+namespace Gerrit.Server
+{
+    public class GerritCapabilities : Translate
+    {
+        #region Translation
+        private static readonly TranslationString _publishTypeReview = new TranslationString("For Review");
+        private static readonly TranslationString _publishTypeWip = new TranslationString("Work-in-Progress");
+        private static readonly TranslationString _publishTypePrivate = new TranslationString("Private");
+        private static readonly TranslationString _publishTypeDraft = new TranslationString("Draft");
+        #endregion
```
Comment (dfev77) [inline Plugins/Gerrit/Server/GerritCapabilities.cs]: I have moved these from FormGerritPublish.cs but UpdateLocalEnglishTranslations script did not add them to the xlf file.
How to handle this?

**Comment by RussKie** [comment]
Thank you.

If you able to send your original simple fix to release/3.3 branch in the next few days, I will include it in 3.3.1.
